velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/aggregates/BloomFilterAggAggregate.h"
+
+#include "velox/common/base/BloomFilter.h"
+#include "velox/exec/Aggregate.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::functions::sparksql::aggregates {
+
+namespace {
+
+struct BloomFilterAccumulator {
+  explicit BloomFilterAccumulator(HashStringAllocator* allocator)
+      : bloomFilter_{StlAllocator<uint64_t>(allocator)} {}
+
+  int32_t serializedSize() {
+    return bloomFilter_.serializedSize();
+  }
+
+  void serialize(char* output) {
+    return bloomFilter_.serialize(output);
+  }
+
+  void mergeWith(StringView& serialized) {
+    bloomFilter_.merge(serialized.data());
+  }
+
+  void init(int32_t capacity) {
+    if (!bloomFilter_.isSet()) {
+      bloomFilter_.reset(capacity);
+    }
+  }
+
+  void insert(int64_t value) {
+    bloomFilter_.insert(folly::hasher<int64_t>()(value));
+  }
+
+  BloomFilter<StlAllocator<uint64_t>> bloomFilter_;
+}; // namespace
+
+class BloomFilterAggAggregate : public exec::Aggregate {
+ public:
+  explicit BloomFilterAggAggregate(const TypePtr& resultType)
+      : Aggregate(resultType) {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(BloomFilterAccumulator);
+  }
+
+  bool isFixedSize() const override {
+    return false;
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    setAllNulls(groups, indices);
+    for (auto i : indices) {
+      new (groups[i] + offset_) BloomFilterAccumulator(allocator_);
+    }
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    decodeArguments(rows, args);
+    VELOX_USER_CHECK(!decodedRaw_.mayHaveNulls());
```

Comment:
Spark bloomfilter aggregate test only tests the empty input. https://github.com/apache/spark/blob/branch-3.3/sql/core/src/test/scala/org/apache/spark/sql/BloomFilterAggregateQuerySuite.scala#L196
I add the test: 
```
  test("Test that bloom_filter_agg errors null") {
    spark.sql(
      """
        |SELECT bloom_filter_agg(null)"""
        .stripMargin)
  }
```
It will throw exception:
```
[DATATYPE_MISMATCH.BLOOM_FILTER_WRONG_TYPE] Cannot resolve "bloom_filter_agg(NULL, 1000000, 8388608)" due to data type mismatch: Input to function `bloom_filter_agg` should have been "BINARY" followed by value with "BIGINT", but it's ["VOID", "BIGINT", "BIGINT"].; line 2 pos 7;
```
In spark, its first argument is xxhash(table_col), so it won't be null.

Velox BloomFilter accepts uint64_t while `xxhash()` returns int64_t, so we need to use folly to hash twice.

Reply:
> In spark, its first argument is xxhash(table_col), so it won't be null.

I think it is totally possible that table_col is null for some or all rows.

> SELECT bloom_filter_agg(null)

Try changing this to something like this:

```
SELECT bloom_filter_agg(cast(null as varbinary))
```

Reply:
I change the test to
```
  test("Test that bloom_filter_agg errors null") {
    spark.sql(
      """
        |SELECT bloom_filter_agg(cast(null as binary))"""
        .stripMargin)
  }
```

Different exception:
```
[DATATYPE_MISMATCH.BLOOM_FILTER_WRONG_TYPE] Cannot resolve "bloom_filter_agg(CAST(NULL AS BINARY), 1000000, 8388608)" due to data type mismatch: Input to function `bloom_filter_agg` should have been "BINARY" followed by value with "BIGINT", but it's ["BINARY", "BIGINT", "BIGINT"].; line 2 pos 7;
'Aggregate [unresolvedalias(bloom_filter_agg(cast(null as binary), 1000000, 8388608, 0, 0), None)]
+- OneRowRelation
```

This is spark internal function, it is revoked by the planner, for the case `table_col is null`, it will be
`bloom_filter_agg(xxhash64(null))`, and xxhash64(null) is 42

Reply:
> This is spark internal function, it is revoked by the planner, for the case table_col is null, it will be
bloom_filter_agg(xxhash64(null)), and xxhash64(null) is 42

Interesting. So the input to bloom_filter_agg is not a value, but a hash of the value. Let's clarify this in the documentation. What's the type if input? Is it VARBINARY or BIGINT?

Reply:
```
 val rowCount = filterCreationSidePlan.stats.rowCount
    val bloomFilterAgg = new BloomFilterAggregate(new XxHash64(Seq(filterCreationSideExp)), rowCount.get.longValue)
```

It is BIGINT